### PR TITLE
Use ZkCacheBaseDataAccessor to cache instance configs in PinotHelixResourceManager

### PR DESCRIPTION
**Pull request by jackjlli**
This PR uses `ZkCacheBaseDataAccessor` to cache instance configs in PinotHelixResourceManager, so that zk read access can be reduced.

### CODE REVIEW

File: pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/PinotHelixResourceManager.java [hunk 1]
```diff
@@ -276,6 +295,11 @@ public Builder getKeyBuilder() {
    */
   @Nonnull
   public InstanceConfig getHelixInstanceConfig(@Nonnull String instanceId) {
+    if (_cacheInstanceConfigsDataAccessor != null) {
+      LOGGER.info("Get instance config for instance {} from cluster {}.", instanceId, _helixClusterName);
```
Comment (mcvsubbu) [inline pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/PinotHelixResourceManager.java]: Change this to debug log, or remove it altogether please

File: pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/PinotHelixResourceManager.java [hunk 2]
```diff
@@ -261,6 +276,10 @@ public Builder getKeyBuilder() {
     return _keyBuilder;
   }
 
+  public ZkCacheBaseDataAccessor<ZNRecord> getCacheInstanceConfigsDataAccessor() {
```
Comment (mcvsubbu) [inline pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/PinotHelixResourceManager.java]: why do we need this method?

File: pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/PinotHelixResourceManager.java [hunk 3]
```diff
@@ -276,6 +295,11 @@ public Builder getKeyBuilder() {
    */
   @Nonnull
   public InstanceConfig getHelixInstanceConfig(@Nonnull String instanceId) {
+    if (_cacheInstanceConfigsDataAccessor != null) {
```
Comment (mcvsubbu) [inline pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/PinotHelixResourceManager.java]: Call this method in your test

File: pinot-controller/src/test/java/com/linkedin/pinot/controller/helix/PinotResourceManagerTest.java [hunk 4]
```diff
@@ -66,6 +69,13 @@ public void setUp() throws Exception {
     Assert.assertEquals(_helixAdmin.getInstancesInClusterWithTag(HELIX_CLUSTER_NAME, "DefaultTenant_REALTIME").size(),
         1);
 
+    // Verifying instance configs.
+    ZNRecord znRecord = _pinotHelixResourceManager.getCacheInstanceConfigsDataAccessor().get("/Server_localhost_0", null, AccessOption.PERSISTENT);
```
Comment (mcvsubbu) [inline pinot-controller/src/test/java/com/linkedin/pinot/controller/helix/PinotResourceManagerTest.java]: You should invoke the other method in the test, and compare the result against getting the instanceconfig directly from helix.

File: pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/PinotHelixResourceManager.java [hunk 5]
```diff
@@ -168,6 +173,16 @@ public synchronized void start() {
     _helixAdmin = _helixZkManager.getClusterManagmentTool();
     _propertyStore = _helixZkManager.getHelixPropertyStore();
     _helixDataAccessor = _helixZkManager.getHelixDataAccessor();
+    // Cache instance zk paths.
+    BaseDataAccessor<ZNRecord> baseDataAccessor = _helixDataAccessor.getBaseDataAccessor();
+    if (baseDataAccessor instanceof ZkBaseDataAccessor) {
+      String instanceConfigs = PropertyPathBuilder.instanceConfig(_helixClusterName);
+      _cacheInstanceConfigsDataAccessor =
+          new ZkCacheBaseDataAccessor<>((ZkBaseDataAccessor<ZNRecord>) baseDataAccessor, instanceConfigs,
+              Collections.singletonList(instanceConfigs), Collections.singletonList(instanceConfigs));
```
Comment (dasahcc) [inline pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/PinotHelixResourceManager.java]: If you already give root path. You should not put INSTANCE path again here.  Or if you know that instances are stable, which means there are no instances added or shutdown, you can directly subscribe the paths without root path.

You can refer Helix ZkHelixPropertyStore or this constructor:

  public ZkCacheBaseDataAccessor(ZkBaseDataAccessor<T> baseAccessor, List<String> wtCachePaths) {
    this(baseAccessor, null, wtCachePaths, null);
  }

Comment (jackjlli) [reply pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/PinotHelixResourceManager.java]: @dasahcc Do you mean keeping one `Collections.singletonList(instanceConfigs)` is enough if I've already given root path?
We can't guarantee instances are stable; we may add/remove instances. But the purpose of getting instance configs is to get the admin endpoint, which consists of hostname and admin port. And these two elements won't be changed often for each instance. So I guess the correct usage of that class is:
```
_cacheInstanceConfigsDataAccessor =
          new ZkCacheBaseDataAccessor<>((ZkBaseDataAccessor<ZNRecord>) baseDataAccessor, instanceConfigs,
              Collections.singletonList(instanceConfigs), null);
```
Could you confirm?

Comment (dasahcc) [reply pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/PinotHelixResourceManager.java]: That should work. If possible, you can local start a test and modify the instance config through zooinspector to see whether it got updated at cache side.

Comment (jackjlli) [reply pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/PinotHelixResourceManager.java]: Great! I've tested it using your suggestion and it works perfectly. Thanks for the help! 👍

File: pinot-controller/src/test/java/com/linkedin/pinot/controller/helix/core/PinotHelixResourceManagerTest.java [hunk 6]
```diff
@@ -76,6 +83,31 @@ public void testGetInstanceEndpoints() throws InvalidConfigException {
...
+        InstanceConfig instanceConfig = new InstanceConfig(znRecord);
+        String[] hostnameSplit = instanceConfig.getHostName().split("_");
+        String adminEndpoint =  hostnameSplit[1] + ":" + adminPort;
+        Assert.assertTrue(endpoints.inverse().containsKey(adminEndpoint));
+      }
+    }
+  }
+
+  @Test
+  public void testGetInstanceConfigsFromZkCacheBaseDataAccessor() {
```
Comment (mcvsubbu) [inline pinot-controller/src/test/java/com/linkedin/pinot/controller/helix/core/PinotHelixResourceManagerTest.java]: Isnt this test a duplicate of lines 86 through 97?

Also, can u add a test (or tests) where a new instance is added and an old one removed, and the cached data accessor shows the modifiction correctly? I think it is worth the effort.

File: pinot-controller/src/test/java/com/linkedin/pinot/controller/helix/core/PinotHelixResourceManagerTest.java [hunk 7]
```diff
@@ -76,6 +86,60 @@ public void testGetInstanceEndpoints() throws InvalidConfigException {
...
+        InstanceConfig instanceConfig = new InstanceConfig(znRecord);
+        String[] hostnameSplit = instanceConfig.getHostName().split("_");
+        String adminEndpoint =  hostnameSplit[1] + ":" + adminPort;
+        Assert.assertTrue(endpoints.inverse().containsKey(adminEndpoint));
+      }
+    }
+  }
+
+  @Test
+  public void testGetInstanceConfigsFromZkCacheBaseDataAccessor() throws InterruptedException {
```
Comment (mcvsubbu) [inline pinot-controller/src/test/java/com/linkedin/pinot/controller/helix/core/PinotHelixResourceManagerTest.java]: change name of test

File: pinot-controller/src/test/java/com/linkedin/pinot/controller/helix/core/PinotHelixResourceManagerTest.java [hunk 8]
```diff
@@ -76,6 +86,60 @@ public void testGetInstanceEndpoints() throws InvalidConfigException {
...
+    String instanceConfigPath = PropertyPathBuilder.instanceConfig(_helixClusterName, instanceName);
+    Assert.assertTrue(zkClient.exists(instanceConfigPath));
+    ZNRecord znRecord = zkClient.readData(instanceConfigPath, null);
+
+    InstanceConfig cachedInstanceConfig = _helixResourceManager.getHelixInstanceConfig(instanceName);
+    String originalPort = cachedInstanceConfig.getPort();
+    Assert.assertNotNull(originalPort);
+    String newPort = Long.toString(System.currentTimeMillis());
+    Assert.assertTrue(!newPort.equals(originalPort));
+
```
Comment (mcvsubbu) [inline pinot-controller/src/test/java/com/linkedin/pinot/controller/helix/core/PinotHelixResourceManagerTest.java]: also add/delete instances and wait for the change to show up

File: pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/PinotHelixResourceManager.java [hunk 9]
```diff
@@ -276,7 +292,8 @@ public Builder getKeyBuilder() {
    */
   @Nonnull
   public InstanceConfig getHelixInstanceConfig(@Nonnull String instanceId) {
-    return _helixAdmin.getInstanceConfig(_helixClusterName, instanceId);
+    ZNRecord znRecord = _cacheInstanceConfigsDataAccessor.get("/" + instanceId, null, AccessOption.PERSISTENT);
```
Comment (mcvsubbu) [inline pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/PinotHelixResourceManager.java]: pls change the other methods to get instanceconfigs to use the cached handle. thanks,

**Comment by npawar** [comment]
Will you also be changing `public List<InstanceConfig> getAllHelixInstanceConfigs()` method to read from cache? Currently it is still reading from zk right?

**Comment by jackjlli** [comment]
@npawar Yes, the method `getAllHelixInstanceConfigs()` also uses cached data accessor.

File: pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/PinotHelixResourceManager.java (hunk 9 above)
Comment (jackjlli) [reply pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/PinotHelixResourceManager.java]: Done

File: pinot-controller/src/test/java/com/linkedin/pinot/controller/helix/core/PinotHelixResourceManagerTest.java (hunk 8 above)
Comment (jackjlli) [reply pinot-controller/src/test/java/com/linkedin/pinot/controller/helix/core/PinotHelixResourceManagerTest.java]: Another tests added.

**Pull request merged**